kumascript/macros/CSSSyntax.ejs

```diff
@@ -1,442 +1,324 @@
-                        typesSyntax += "<br/>";
-                    }
-                    describedTypes.push(type);
-                }
-            }
-
-            if (typesSyntax !== "") {
-                formattedTypesSyntax += "<p style=\"font-family:Open Sans,Arial," +
-                    "sans-serif; margin: 10px 0 0 0;\">" + s_where +
-                    "<br/><code style=\"font-family:Consolas,Monaco," +
-                    "&quot;Andale Mono&quot;,monospace;\">" + typesSyntax +
-                    "</code></p>";
-            }
-        }
-
-        return formattedTypesSyntax + await formatTypesSyntax(typesSyntax, describedTypes);
-    }
+/**
+ * Render the syntax for a single type.
+ */
+function renderSyntax(type, syntax) {
+  // write out the name of this type
+  let output = `<span class="token property" id="${type}">${type} = </span><br/>`;
+
+  const ast = definitionSyntax.parse(syntax);
+  // if the combinator is ' ', write the complete type syntax in a single line
+  if (ast.combinator === ' ') {
+    output += renderTerms([ast], ast.combinator);
+  } else {
+  // otherwise write out each direct child in its own line
+    output += renderTerms(ast.terms, ast.combinator);
+  }
 
-    return formattedTypesSyntax;
+  return output;
 }
 
-if (!atRule) {
-    var matches = null;
-    if (slug) {
-        matches = slug.match(/\/CSS\/(@.+?)(?=\/)/);
+/**
+ * Get names of all the types in a given set of syntaxes
+ */
+function getTypesForSyntaxes(syntaxes, constituents) {
+
+  function processNode(node) {
+    if (node.type === 'Type' &&
+       (constituents.indexOf(node.name) === -1)) {
+      constituents.push(node.name);
     }
+  }
+
+  for (const syntax of syntaxes) {
+    let ast = definitionSyntax.parse(syntax);
+    definitionSyntax.walk(ast, processNode);
+  }
+
+}
 
-    if (matches) {
-        atRule = matches[1];
+/**
+ * Given an item (such as a CSS property), fetch all the types that participate
+ * in its formal syntax definition, either directly or transitively.
+ */
+function getConstituentTypes(propertySyntax) {
+  const allConstituents = [];
+  let oldConstituentsLength = allConstituents.length;
+  // get all the types in the top-level syntax
+  let constituentSyntaxes = [propertySyntax];
+  getTypesForSyntaxes(constituentSyntaxes, allConstituents);
+
+  // while an iteration added more types...
+  while (allConstituents.length > oldConstituentsLength) {
+    // get the syntaxes for all newly added constituents,
+    // and then get the types in those syntaxes
+    constituentSyntaxes = [];
+    for (let constituent of allConstituents.slice(oldConstituentsLength)) {
+
+      let constituentSyntaxEntry = valuespaces[`<${constituent}>`];
+
+      if (constituentSyntaxEntry && constituentSyntaxEntry.value) {
+        constituentSyntaxes.push(constituentSyntaxEntry.value);
+      }
     }
+
+    oldConstituentsLength = allConstituents.length;
+    getTypesForSyntaxes(constituentSyntaxes, allConstituents);
```

Comment:
Those duplicated statements in lines 259/262 and 278/279 should be merged, e.g. like this:

```suggestion
  let oldConstituentsLength = 0;
  // get all the types in the top-level syntax
  let constituentSyntaxes = [propertySyntax];

  // while an iteration added more types...
  while (true) {
    oldConstituentsLength = allConstituents.length;
    getTypesForSyntaxes(constituentSyntaxes, allConstituents);
  
    if (allConstituents.length <= oldConstituentsLength) {
      break;
    }
    // get the syntaxes for all newly added constituents,
    // and then get the types in those syntaxes
    constituentSyntaxes = [];
    for (let constituent of allConstituents.slice(oldConstituentsLength)) {

      let constituentSyntaxEntry = valuespaces[`<${constituent}>`];

      if (constituentSyntaxEntry && constituentSyntaxEntry.value) {
        constituentSyntaxes.push(constituentSyntaxEntry.value);
      }
    }
```